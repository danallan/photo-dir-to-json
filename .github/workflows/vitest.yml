name: Run tests with coverage, build code, and build docs
on:
  push:
  pull_request:
    branches: [main]

jobs:
  Test-and-Build:
    timeout-minutes: 60
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: 'Install node'
        uses: actions/setup-node@v4
        with:
          node-version: 20
      - name: 'Install dependencies'
        run: |
          corepack enable
          yarn install --immutable
      - name: 'Run build'
        run: yarn build
      - name: 'Report Coverage'
        # https://github.com/marketplace/actions/vitest-coverage-report
        if: always()
        uses:  davelosert/vitest-coverage-report-action@v2