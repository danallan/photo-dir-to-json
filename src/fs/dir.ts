import { Dirent, readdirSync } from 'fs';
import { basename, join as pathJoin, resolve } from 'path';

/**
 * A set of helper functions for a local directory.
 * @internal
 */
export class Directory {
    private _subdirs: null | Directory[] = null;
    private _name: string;
    readonly path;

    constructor(path: string) {
        this.path = resolve(path);
        this._name = basename(path);
    }

    get name() {
        return this._name;
    }

    private contents(): Dirent[] {
        return readdirSync(this.path, { withFileTypes: true });
    }

    public getSubDirectoriesSync(except: string[]): Directory[] {
        if (this._subdirs === null) {
            this._subdirs = this.contents()
                                .filter(f => f.isDirectory())
                                .filter(f => !except.includes(f.name))
                                .map(f => pathJoin(f.path, f.name))
                                .map(p => new Directory(p));
        }
        return this._subdirs;
    }

    public getFilesSync(): string[] {
        return this.contents()
            .filter(f => f.isFile())
            .map(f => f.name);
    }
}