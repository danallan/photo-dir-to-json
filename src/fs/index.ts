export { Directory } from './dir.js';