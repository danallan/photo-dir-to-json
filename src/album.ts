import { Directory } from './fs';
import { Photo } from './photo';
import {
    albumSchemaType,
    photoSchemaType,
    metadataSchema,
    metadataSchemaType
} from './schema';
import { existsSync, readFileSync, writeFileSync } from 'fs';
import { join as pathJoin, extname } from 'path';

/**
 * Optional album configuration.
 * @public
*/
export interface AlbumOptions {
    /** The path to a directory containing album metadata file.
      * If this is specified, the metadata file is assumed to match the album's
      * directory name. This cannot be specified with `metadataFile`.
      */
    metadataDir?: string,
    /** The name of the metadata file inside of the album directory. E.g.,
      * `_metadata.json`. This cannot be specified with `metadataDir`.
      */
    metadataFile?: string,
    /** A list of case-insensitive extensions to match photo files inside the
      * album. The files must be supported by
      * {@link https://github.com/mattiasw/ExifReader#readme | ExifReader}
      * library. Default: ['jpg', 'jpeg', 'png', 'webp']
      */
    allowedExtensions?: string[],
    /** Suppress warnings for files that are skipped in the album with these
      * case-insensitive extensions. Default: ['json', 'ds_store']
      */
    skippedExtensions?: string[],
}

/**
 * The Album class represents a local directory of photos and an optional input
 * metadata JSON file that conforms to {@link metadataSchema}. The class
 * collates information about each photo inside of the album directory, combines
 * it with the (optional but recommended) input metadata and provides it all in
 * a single {@link albumSchema} metadata object.
 * @public
 */
export class Album {
    private _allowedExtensions: Set<string>;
    private _skippedExtensions: Set<string>;
    private _metadata: metadataSchemaType;
    private _photos: Photo[] | null = null;
    private _directory: Directory;

    /**
     *
     * @param path - The on-disk path to the album
     * @param options - Optionally specify the location of album metadata file
     *   and override default extensions used for including photos
     */
    constructor(path: string, options?: AlbumOptions) {
        this._directory = new Directory(path);

        const opts = options ?? {};

        // prefix all extensions with a period to properly compare againt
        // result of path.extname()
        const allow = opts.allowedExtensions ?? ['jpg', 'jpeg', 'png', 'webp'];
        this._allowedExtensions = new Set(allow.map(e => `.${e.toLowerCase()}`));

        const skip = opts.skippedExtensions ?? ['json', 'ds_store'];
        this._skippedExtensions = new Set(skip.map(e => `.${e.toLowerCase()}`));

        if (opts.metadataDir && opts.metadataFile)
            throw new Error("Error: cannot specify both `metadataDir` and `metadataFile` in AlbumOptions");

        // try to load metadata from specified place, if unspecified set default
        let metadataFile: string|null = null;
        if (opts.metadataDir)
            metadataFile = pathJoin(opts.metadataDir, `${this.name}.json`);
        if (opts.metadataFile)
            metadataFile = pathJoin(this.path, opts.metadataFile);

        this._metadata = (metadataFile)
            ? this._loadMetadata(metadataFile)
            : { title: this.name };

        // verify the thumb exists in this album
        if (this._metadata.thumb && !existsSync(pathJoin(this.path, this._metadata.thumb)))
            throw new Error(`Thumb ${this._metadata.thumb} not found in album ${this.name}`);
    }

    /**
     *
     * @param file - full path to album metadata json file
     * @returns parsed metadata
     */
    private _loadMetadata(file: string): metadataSchemaType {
        const contents = readFileSync(file, 'utf8');
        return metadataSchema.parse(JSON.parse(contents));
    }

    /**
     *
     * @param filename - full path to photo file on disk
     * @returns true if filename extension is in allowedExtensions, false
     *   otherwise
     */
    private _selectOnlyPhotos(filename: string) {
        const extension = extname(filename).toLowerCase();
        if (this._allowedExtensions.has(extension))
            return true;
        if (!this._skippedExtensions.has(extension))
            console.error(`WARNING: skipping file ${filename} in album ${this.name}`);
        return false;
    }

    /**
     * List of Photo classes for all images found in the album path. Only files
     * with extensions from {@link AlbumOptions | `allowedExtensions` property}
     * are processed, all others are ignored. Any ignored file extensions not
     * specified in {@link AlbumOptions | `skippedExtensions` property} prints a
     * warning to the console notifying you that they were skipped.
     * @returns List of Photo instances for all valid photos in the album
     */
    get photos(): Photo[] {
        if (this._photos === null) {
            this._photos = this._directory.getFilesSync()
                .filter(f => this._selectOnlyPhotos(f))
                .map(f => new Photo(this.path, f));
        }
        return this._photos;
    }

    /**
     * By default, album title is the album directory name but is overridden by
     * the `title` property in album metadata when supplied.
     * @returns the album title
     */
    get title(): string {
        return this._metadata.title;
    }

    /**
     * On-disk path to the album, e.g. `/Volume/Photos/Album1`.
     * @returns string path
     */
    get path(): string {
        return this._directory.path;
    }

    /**
     * The `slug` field defined in the Album metadata, defaulting to the Album's
     * on-disk directory name lower cased. For example, if the slug is not
     * specified in the input metadata, the slug for an album located at
     * `/Volume/Photos/Album1` becomes `album1`.
     */
    get slug(): string {
        return this._metadata.slug || this.name.toLowerCase();
    }

    /**
     * The on-disk name of the album's directory. If the path is
     * `/Volume/Photos/Album1` the name is `Album1`.
     * @returns directory name of the album
     */
    get name(): string {
        return this._directory.name;
    }

    private _albumMetadata(photos: photoSchemaType[]): albumSchemaType {
        return {
            ...this._metadata,
            title: this.title,
            slug: this.slug,
            unlisted: this._metadata.unlisted === true,
            photos,
        }
    }

    /**
     * Returns the full metadata for the album including all photos inside.
     * @returns A promise that resolves to full album metadata after photo
     *   metadata is read from disk
     * @public
     */
    public async metadata(): Promise<albumSchemaType> {
        let photos: photoSchemaType[] = [];
        for (const photo of this.photos) {
            photos.push(await photo.metadata());
        }
        return this._albumMetadata(photos);
    }

    /**
     * Write the Album's generated metadata to a JSON file on disk. Please
     * ensure the directories exist and the caller has permissions to write to
     * the path. The file at the specified path will be overwritten if it
     * exists.
     * @param file - full file path of JSON file to write
     */
    public async saveMetadata(file: string) {
        const out = await this.metadata();
        writeFileSync(file, JSON.stringify(out));
    }
}