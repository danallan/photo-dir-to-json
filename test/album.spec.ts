import { describe, test, vi } from 'vitest';
import { Album } from '../src/album.js';
import { albumSchema } from '../src/schema/index.js';
import { resolve, join as pathJoin } from 'path';
import { readFileSync, rmSync } from 'fs';

// all test images
const images = './test/Images';

// a narrowly defined album for snapshots
const albumdir = './test/Album'
const metadataDir = './test/metadata';
const metadataFile = '_metadata.json';

describe('getters', () => {
    test('name is basename', ({ expect }) => {
        const album = new Album(images);
        expect(album.name).toEqual('Images');
    });

    test('path returns absolute path', ({ expect }) => {
        const album = new Album(albumdir);
        const path = resolve(albumdir);
        expect(album.path).toEqual(path);
    });

    test('slug is lowercase name by default', ({ expect }) => {
        const album = new Album(albumdir);
        expect(album.slug).toEqual('album');
    });

    test('title equals name by default', ({ expect }) => {
        const album = new Album(albumdir);
        expect(album.title).toEqual(album.name);
    });

    test('photos includes all jpg, jpeg, png, webp, no json', ({ expect }) => {
        const photos = (new Album(images)).photos;
        expect(photos.length).toBe(13);

        const names = photos.map(p => p.name);
        expect(names).toContain('exif.jpg');
        expect(names).toContain('exif.png');
        expect(names).toContain('exif.webp');
    });

    test('photos can be empty', ({ expect }) => {
        const photos = (new Album(metadataDir)).photos;
        expect(photos).toEqual([]);
    });
});

describe('options object', () => {
    test('cannot specify both metadataDir and metadataFile', ({ expect }) => {
        const opts = { metadataDir, metadataFile };
        expect(() => { new Album(albumdir, opts); }).toThrow();
    });

    test('allowedExtensions overrides default', ({ expect }) => {
        const consoleSpy = vi.spyOn(console, 'error');
        expect(consoleSpy.mock.calls.length).toBe(0);

        const opts = { allowedExtensions: ['png'] };
        const photos = (new Album(images, opts)).photos;
        const names = photos.map(p => p.name);

        // all PNG files
        expect(names).toEqual(['exif.png', 'wide.png']);

        // should also trigger console errors for unskipped files
        // don't forget that _metadata.json by default does not emit error
        expect(consoleSpy).toHaveBeenCalledTimes(11);

        consoleSpy.mockRestore();
    });

    test('skippedExtensions overrides default', ({ expect }) => {
        const opts = {
            allowedExtensions: [],
            skippedExtensions: ['json', 'jpg', 'png', 'webp'],
        };
        const consoleSpy = vi.spyOn(console, 'error');
        (new Album(images, opts)).photos;

        // expect no warnings
        expect(consoleSpy).not.toBeCalled();

        consoleSpy.mockRestore();
    });
});

describe('metadata', () => {
    test('title getter is overwritten by metadata', ({ expect }) => {
        const album = new Album(albumdir, { metadataFile });
        expect(album.title).toBe('Test album');
    });

    test('slug getter is overwritten by metadata', ({ expect }) => {
        const album = new Album(albumdir, { metadataFile });
        expect(album.slug).toBe('/path/and-a-slug-title');
    });

    test('throws error when thumb specified but file not found', ({ expect }) => {
        const opts = { metadataFile: '_metadata-missing_title.json' };
        expect(() => { new Album(albumdir, opts) }).toThrow(/Required/);
    });

    test('throws error when thumb specified but file not found', ({ expect }) => {
        const opts = { metadataFile: '_metadata-missing_thumb.json' };
        expect(() => { new Album(albumdir, opts) }).toThrow(/Thumb not found/);
    });

    test('throws error when metadata file has extra property', ({ expect }) => {
        const opts = { metadataFile: '_metadata-strict.json' };
        expect(() => { new Album(albumdir, opts) }).toThrow(/unrecognized_keys/)
    });

    test('read from metadataDir', ({ expect }) => {
        const album = new Album(albumdir, { metadataDir });
        expect(album.slug).toBe('metadata-from-dir');
    });

    test('full from file matches snapshot', async ({ expect }) => {
        const album = new Album(albumdir, { metadataFile });
        const metadata = await album.metadata();
        expect(metadata).toMatchSnapshot();
    });

    test('full from dir matches snapshot', async ({ expect }) => {
        const album = new Album(albumdir, { metadataDir });
        const metadata = await album.metadata();
        expect(metadata).toMatchSnapshot();
    });
});

test('saveMetadata() writes valid JSON', async ({ expect }) => {
    const album = new Album(albumdir, { metadataFile });
    const json = pathJoin(metadataDir, 'out.json');
    await album.saveMetadata(json);

    // now read it back, shouldn't throw
    const contents = readFileSync(json, 'utf8');
    const written = albumSchema.parse(JSON.parse(contents));

    // sanity check
    expect(written.title).toBe(album.title);
    expect(written.photos.length).toBe(1);

    // cleanup
    rmSync(json);
});