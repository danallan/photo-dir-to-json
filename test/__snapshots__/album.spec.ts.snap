// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`metadata > full from dir matches snapshot 1`] = `
{
  "description": "metadata dir",
  "keywords": [
    "no photos here",
    "test",
  ],
  "order": [
    "IMG_1234.jpg",
  ],
  "photos": [
    {
      "date": "2023-01-01T08:00:01.000Z",
      "filename": "IMG_1234.jpg",
      "height": 16,
      "landscape": false,
      "width": 16,
    },
  ],
  "slug": "metadata-from-dir",
  "thumb": "IMG_1234.jpg",
  "title": "Test album",
  "unlisted": false,
}
`;

exports[`metadata > full from file matches snapshot 1`] = `
{
  "description": "metadata file",
  "keywords": [
    "no photos here",
    "test",
  ],
  "order": [
    "IMG_1234.jpg",
  ],
  "photos": [
    {
      "date": "2023-01-01T08:00:01.000Z",
      "filename": "IMG_1234.jpg",
      "height": 16,
      "landscape": false,
      "width": 16,
    },
  ],
  "slug": "/path/and-a-slug-title",
  "thumb": "IMG_1234.jpg",
  "title": "Test album",
  "unlisted": true,
}
`;
